### ISSUE TITLE
LHS of assignment expressions are never(?) parsed as AssignmentPatterns

### ISSUE THREAD

**Issue by caitp** [google/traceur-compiler#1833]
The following is reported as a syntax error:

```js
var a, b;
({a, b}) = {a: 1, b: 2};
```

Based on the draft and other implementations, it seems like this is not right:

From the draft:

>In certain circumstances when processing the production AssignmentExpression : LeftHandSideExpression = AssignmentExpression the following grammar is used to refine the interpretation of LeftHandSideExpression.

(particularly, when the LHS looks like an object literal or array literal)

From SpiderMonkey:

```c++
/*
 * Destructuring patterns can appear in two kinds of contexts:
 *
 * - assignment-like: assignment expressions and |for| loop heads.  In
 *   these cases, the patterns' property value positions can be
 *   arbitrary lvalue expressions; the destructuring is just a fancy
 *   assignment.
 *
 * - binding-like: |var| and |let| declarations, functions' formal
 *   parameter lists, |catch| clauses, and comprehension tails.  In
 *   these cases, the patterns' property value positions must be
 *   simple names; the destructuring defines them as new variables.
 *
 * In both cases, other code parses the pattern as an arbitrary
 * primaryExpr, and then, here in checkDestructuring, verify that the
 * tree is a valid AssignmentPattern or BindingPattern.
 *
 * In assignment-like contexts, we parse the pattern with
 * pc->inDeclDestructuring clear, so the lvalue expressions in the
 * pattern are parsed normally.  primaryExpr links variable references
 * into the appropriate use chains; creates placeholder definitions;
 * and so on.  checkDestructuring is called with |data| nullptr (since
 * we won't be binding any new names), and we specialize lvalues as
 * appropriate.
 *
 * In declaration-like contexts, the normal variable reference
 * processing would just be an obstruction, because we're going to
 * define the names that appear in the property value positions as new
 * variables anyway.  In this case, we parse the pattern with
 * pc->inDeclDestructuring set, which directs primaryExpr to leave
 * whatever name nodes it creates unconnected.  Then, here in
 * checkDestructuring, we require the pattern's property value
 * positions to be simple names, and define them as appropriate to the
 * context.  For these calls, |data| points to the right sort of
 * BindData.
 */
```

(originally reported in https://github.com/jshint/jshint/issues/2269)

**Comment by arv** [comment]
I remember SpiderMonkey doing this but it was never accepted into the spec. I don't think this should be allowed.

I'll have to look closer at the spec again.

**Comment by arv** [comment]
This is valid in SpiderMonkey but it is not valid in ES6.

**Comment by caitp** [comment]
```
AssignmentExpression[In, Yield] : LeftHandSideExpression[?Yield] = AssignmentExpression[?In, ?Yield]

1. If LeftHandSideExpression is neither an ObjectLiteral nor an ArrayLiteral, then
    a. Let lref be the result of evaluating LeftHandSideExpression.
    b. ReturnIfAbrupt(lref).
    c. Let rref be the result of evaluating AssignmentExpression.
    d. Let rval be GetValue(rref).
    e. If IsAnonymousFunctionDefinition(AssignmentExpression) and IsIdentifierRef of LeftHandSideExpression are both true, then
        i. Let hasNameProperty be HasOwnProperty(rval, "name").
        ii. ReturnIfAbrupt(hasNameProperty).
        iii. If hasNameProperty is false, perform SetFunctionName(rval, GetReferencedName(lref)).
    f. Let status be PutValue(lref, rval).
    g. ReturnIfAbrupt(status).
    h. Return rval.
2. Let assignmentPattern be the parse of the source text corresponding to LeftHandSideExpression using AssignmentPattern[?Yield] as the goal symbol.
3. Let rref be the result of evaluating AssignmentExpression.
4. Let rval be GetValue(rref).
5. ReturnIfAbrupt(rval).
6. Let status be the result of performing DestructuringAssignmentEvaluation of assignmentPattern using rval as the argument.
7. ReturnIfAbrupt(status).
8. Return rval.
```

The best reason I can come up with that this isn't allowed, is this:

```
12.3.1.3 Static Semantics: IsDestructuring

See also: 13.6.4.6.

MemberExpression : PrimaryExpression
    1. If PrimaryExpression is either an ObjectLiteral or an ArrayLiteral, return true.
    2. Return false.
```

since this doesn't seem to account for ParenthesizedExpression/supplemental grammar --- But then, `[a, b] = [1, 2]` should also fail
